optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/bestsolutionmutation/BestSolutionMutationSubSingleStatistic.java

```diff
@@ -21,71 +21,35 @@
         extends ProblemBasedSubSingleStatistic<Solution_, BestSolutionMutationStatisticPoint> {
 
-    private BestSolutionMutationSubSingleStatisticListener listener;
-
     public BestSolutionMutationSubSingleStatistic(SubSingleBenchmarkResult subSingleBenchmarkResult) {
         super(subSingleBenchmarkResult, ProblemStatisticType.BEST_SOLUTION_MUTATION);
-        listener = new BestSolutionMutationSubSingleStatisticListener();
     }
 
     // ************************************************************************
     // Lifecycle methods
     // ************************************************************************
 
     @Override
-    public void open(Solver<Solution_> solver) {
-        DefaultSolver<Solution_> defaultSolver = (DefaultSolver<Solution_>) solver;
-        InnerScoreDirectorFactory<Solution_, ?> innerScoreDirectorFactory = defaultSolver.getScoreDirectorFactory();
-        SolutionDescriptor<Solution_> solutionDescriptor = innerScoreDirectorFactory.getSolutionDescriptor();
-        listener.setMutationCounter(new MutationCounter<>(solutionDescriptor));
-        solver.addEventListener(listener);
+    public void open(StatisticRegistry registry, Tags runTag, Solver<Solution_> solver) {
+        registry.addListener(SolverMetric.BEST_SOLUTION_MUTATION, timestamp -> {
+            registry.getGaugeValue(SolverMetric.BEST_SOLUTION_MUTATION, runTag, mutationCount -> {
+                pointList.add(new BestSolutionMutationStatisticPoint(timestamp, mutationCount.intValue()));
+            });
+        });
     }
 
     @Override
-    public void close(Solver<Solution_> solver) {
-        solver.removeEventListener(listener);
-    }
-
-    private class BestSolutionMutationSubSingleStatisticListener implements SolverEventListener<Solution_> {
-
-        private MutationCounter<Solution_> mutationCounter;
-
-        private Solution_ oldBestSolution = null;
-
-        private void setMutationCounter(MutationCounter<Solution_> mutationCounter) {
-            if (this.mutationCounter != null) {
-                throw new IllegalStateException("Impossible state: mutationCounter (" + this.mutationCounter
-                        + ") is not null.");
-            }
-            this.mutationCounter = mutationCounter;
-        }
-
-        @Override
-        public void bestSolutionChanged(BestSolutionChangedEvent<Solution_> event) {
-            int mutationCount;
-            Solution_ newBestSolution = event.getNewBestSolution();
-            if (oldBestSolution == null) {
-                mutationCount = 0;
-            } else {
-                mutationCount = mutationCounter.countMutations(oldBestSolution, newBestSolution);
-            }
-            pointList.add(new BestSolutionMutationStatisticPoint(
-                    event.getTimeMillisSpent(), mutationCount));
-            oldBestSolution = newBestSolution;
-        }
-
+    public void close(StatisticRegistry registry, Tags runTag, Solver<Solution_> solver) {
```

Comment:
Shouldn't we unregister the listeners here? If so, this needs to happen in all the statistics.
Or is that something that is not supported by micrometer?

Reply:
The `StatisticRegistry` itself is unregistered/removed from `Metrics`. See `SubSingleBenchmarkRunner`:

```
        StatisticRegistry statisticRegistry = new StatisticRegistry((DefaultSolver) solver);
        Metrics.addRegistry(statisticRegistry);
        solver.addEventListener(statisticRegistry);
        solver.addPhaseLifecycleListener(statisticRegistry);

        Tags runTag = Tags.of("optaplanner.benchmark.run", runId);
        for (SubSingleStatistic<Solution_, ?> subSingleStatistic : subSingleBenchmarkResult.getEffectiveSubSingleStatisticMap()
                .values()) {
            subSingleStatistic.open(statisticRegistry, runTag, solver);
            subSingleStatistic.initPointList();
        }
        Solution_ solution = solver.solve(problem);

        solver.removePhaseLifecycleListener(statisticRegistry);
        solver.removeEventListener(statisticRegistry);
        Metrics.removeRegistry(statisticRegistry);
```

Reply:
In that case, I wouldn't override these methods. Make it default, add a comment in there.